File: N-Queens/N-Queens.cpp

class Solution {
public:
    bool ok(vector<int> visit,int endRow)
    {
        for(int row=0;row<endRow;row++)
            if(visit[row]==visit[endRow] || abs(visit[row]-visit[endRow])==endRow-row)
                return false;
        return true;
    }
    void dfs(vector<int> &visit,int n,int row,vector<vector<string> > &ans)
    {
        if(row==n)
        {
            vector<string> vec;
            for(int i=0;i<n;i++)
            {
                string str="";
                for(int j=0;j<n;j++)
                {
                    if(visit[i]==j)
                        str+='Q';
                    else
                        str+='.';
                }
                vec.push_back(str);
            }
            ans.push_back(vec);
            return ;
        }
        for(int col=0;col<n;col++)
        {
            visit[row]=col;
            if(ok(visit,row))
                dfs(visit,n,row+1,ans);
        }
    }
    vector<vector<string> > solveNQueens(int n) {
        vector<vector<string> > ans;
        vector<int> visit(n,0);
        dfs(visit,n,0,ans);
        return ans;
    }
};


 

File: N-Queens/N-Queens.py


class Solution:
    def isok(self,pos,queen):
        for i in range(pos):
            if queen[i]==queen[pos]:
                return False
            if abs(i-pos)==abs(queen[i]-queen[pos]):
                return False
        return True
    def dfs(self,pos,n,queen,ans):
        if pos==n:
            vec=[]
            for i in range(n):
                s=""
                for j in range(n):
                    if queen[i]==j:
                        s+='Q'
                    else:
                        s+='.'
                vec.append(s)
            ans.append(vec)
            return 
        for i in range(n):
            queen[pos]=i
            if self.isok(pos,queen):
                self.dfs(pos+1,n,queen,ans)
    # @return a list of lists of string
    def solveNQueens(self, n):
        queen=[0 for i in range(15)]
        ans=[]
        self.dfs(0,n,queen,ans)
        return ans
 